content/issue-2/how-i-build-backon.md

```diff
@@ -0,0 +1,191 @@
+
+> Once a future has completed (returned `Ready` from `poll`), calling its `poll` method again may panic, block forever, or cause other kinds of problems
+
+Next, I needed to adjust my thinking and focus on implementing for `|| -> impl Future<Result<T>>`. First, I defined a `Retryable` trait and implemented it for all `FnMut() -> Fut` types:
+
+```rust
+pub trait Retryable<
+    B: BackoffBuilder,
+    T,
+    E,
+    Fut: Future<Output = Result<T, E>>,
+    FutureFn: FnMut() -> Fut,
+>
+{
+    /// Generate a new retry
+    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, Fut, FutureFn>;
+}
+
+impl<B, T, E, Fut, FutureFn> Retryable<B, T, E, Fut, FutureFn> for FutureFn
+where
+    B: BackoffBuilder,
+    Fut: Future<Output = Result<T, E>>,
+    FutureFn: FnMut() -> Fut,
+{
+    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, Fut, FutureFn> {
+        Retry::new(self, builder.build())
+    }
+}
+```
+
+This trait involves the following type parameters:
+
+- `B: BackoffBuilder`: the backoff builder passed in by the user, which specifies different backoff parameters
+- `FutureFn: FnMut() -> Fut`: indicates that its type is a function that returns a `Fut`
+  - `FnOnce` requires taking ownership and cannot be called multiple times
+  - `Fn` can only get a `&self` reference, which can be limited in many scenarios
+- `Fut: Future<Output = Result<T, E>>`: represents a Future that returns a `Result<T, E>`
+
+The returned Retry struct wraps all of the above types:
+
+```rust
+pub struct Retry<B: Backoff, T, E, Fut: Future<Output = Result<T, E>>, FutureFn: FnMut() -> Fut> {
+    backoff: B,
+    retryable: fn(&E) -> bool,
+    notify: fn(&E, Duration),
+    future_fn: FutureFn,
+
+    #[pin]
+    state: State<T, E, Fut>,
+}
+```
+
+Apart from `backoff` and `future_fn`, we introduced `retryable` and `notify` to implement retryable error checking and notification functions. Once the type system is clear, the next step is to implement the correct `Future` trait for `Retry`, and the details will not be elaborated:
+
+```rust
+impl<B, T, E, Fut, FutureFn> Future for Retry<B, T, E, Fut, FutureFn>
+where
+    B: Backoff,
+    Fut: Future<Output = Result<T, E>>,
+    FutureFn: FnMut() -> Fut,
+{
+    type Output = Result<T, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ...
+    }
+}
+```
+
+In addition, there are some transactional tasks that need to be completed: we need to let users define which errors can be retried, and provide custom notification for retrying.
+
+Finally, the combined effect is as follows:
+
+```rust
+#[tokio::main]
+async fn main() -> Result<()> {
+    let content = fetch
+        .retry(&ExponentialBuilder::default())
+  		.when(|e| e.to_string() == "EOF")
+        .notify(|err, dur| {
+            println!("retrying error {:?} with sleeping {:?}", err, dur);
```

Comment:
println!("retrying error {:?} with a sleep for {:?}", err, dur); 

 ?? dur is length of sleep?

Reply:
Yes